### PR TITLE
Remove usage of `opcodes::all`

### PR DESCRIPTION
**Pull request by tcharding**
We have all of the opcodes defined in a submodule called `all`, this allows wildcard imports without bringing in the other types in the `opcodes` module.

Use wildcard import `use crate::blockdata::opcodes::all::*` instead of fully qualifying the path to opcodes.

### Original PR description (left here so the thread of discussion below makes sense)

The `all` module adds no value, we can remove it without loss of meaning or clarity, doing so makes the code less verbose.

EDIT: After review, now includes importing with wildcard and removing the `opcodes::` path from any type starting with `OP_`.

Idea stolen from: https://github.com/rust-bitcoin/rust-bitcoin/pull/525 (patch 7)

### CODE REVIEW

**Comment by apoelstra** [comment]
I think we should also drop the `OP_` prefix. Thoughts?

**Comment by tcharding** [comment]
Perhaps rename module to `op` and drop the `OP_` then common usage becomes, for example,`op::VERIFY`.

**Comment by apoelstra** [comment]
Yeah, I like that.

**Comment by tcharding** [comment]
Drats, we can't do it because of things like `OP_2DROP`. We can do the module rename though.

**Comment by apoelstra** [comment]
I'm a little tempted to name `2DROP` etc `_2DROP`. There are only a few of these opcodes and they are practically never used.

@Kixunil what would you think about this? It would be similar to renaming `crate` to `krate` elsewhere in the ecosystem.

**Comment by Kixunil** [comment]
`_` at the beginning means "don't warn if this is unused" so I'd prefer to not do that. Raw identifiers can't begin with numbers either. If renaming `2DROP` to `DROP2` is no-go then we probably need to keep the `OP_` prefix.

**Review by apoelstra — DISMISSED** [review]
ACK e1eb641eacf2e1340b5f4b9135ff8550986aaf33

**Comment by tcharding** [comment]
> We should probably add a commit that adds a deprecated module opcodes and opcodes::all and just re-export everything from both.

By "deprecated" you mean just deprecated to use Rust Bitcoin devs, right? Remembering that I don't currently know of a way to deprecated modules.

**Comment by apoelstra** [comment]
> Remembering that I don't currently know of a way to deprecated modules

Ah, damn, I forgot this.

**Comment by tcharding** [comment]
> know of a way to deprecated modules

That's good English, right there.

**Comment by tcharding** [comment]
Friendly bump, is this one good to go in? Needs a second ACK if so please.

**Review by Kixunil — COMMENTED** [review]
This is moving in the right direction but not sure if we should change this just to change it again if/when we change to enum.

I'm also considering this design to solve #1187:

```rust
/// Types that can be used as opcodes
pub trait Opcode: sealed::Opcode + Into<AnyOpcode> + Copy {
    fn to_u8(self) -> u8;
}

macro_rules! all_opcodes {
    ($($name:ident = $val:expr (, $subclass:ident)*);+ $(;)?) => {
        /// May contain any opcode of bitcoin script.
        #[allow(non_camel_case_types)]
        pub enum AnyOpcode {
            $(
                $name = $val,
            )+
        }
        $(
            /// The zero-sized opcode.
            ///
            /// Zero-sized opcodes can statically guarantee some properties. e.g. (not) carrying bytes parameter. These are represented as conversion traits being implemented.
            #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
            pub struct $name;

            impl From<$name> for AnyOpcode {
                fn from(opcode: $name) -> Self {
                    AnyOpcode::from_u8(opcode.to_u8())
                }
            }

            impl Opcode for $name {
                fn to_u8(self) -> u8 {
                    $value
                }
            }

             impl From<$name> for $subclass {
                fn from(opcode: $name) -> Self {
                    $subclass(opcode.to_u8())
                }
            }
        )+
    }
}

/// Opcodes that don't take a byte array parameter
pub struct NonParametrised(u8);
```

Then the `Instructions` iterator could use `NonParametrised` instead of `All`. `Script::push_opcode` would become: `fn push_opcode<Op: Opcode + Into<NonParametrised>(&mut self, op: Op)`

File: bitcoin/src/address.rs
Comment (apoelstra) [reply bitcoin/src/address.rs:330]: We can't do this because some opcodes start with `2`.

File: bitcoin/src/blockdata/opcodes.rs
Comment (apoelstra) [reply bitcoin/src/blockdata/opcodes.rs]: Because conversion from a u8 to an enum requires either a 256-arm match or a transmute.

**Comment by apoelstra** [comment]
@Kixunil I tend to agree "let's not change this just to change it again". But I remain unconvinced that we can find a much better API than the current messy one.

Maybe if our only goal is to distinguish opcodes which are followed by other non-opcode data, from opcodes which are "zero sized", that would work since this is an unambiguous partition of all the opcodes. But "push vs non-push", "numeric vs non-numeric", cannot be so clearly separated.

This also maybe mixes in with #1324 because we may want some opcodes `CHECKMULTISIG`/`CHECKSIGADD` to only be available in certain contexts. I'm not sure if that'd affect this API.

It all seems like a big difficult problem.

File: bitcoin/src/address.rs
Comment (Kixunil) [reply bitcoin/src/address.rs:330]: I'm suggesting the other change - keep `OP_` and drop `op::`. Now we have `op::OP_` which is kinda silly.

File: bitcoin/src/blockdata/opcodes.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/opcodes.rs]: Neither of them sounds too bad to me.

**Comment by Kixunil** [comment]
Is your idea to just merge it nd iterate on that? That's not too bad for something you doubt can be done (I think it can).

>  But "push vs non-push", "numeric vs non-numeric", cannot be so clearly separated.

My design idea has no problem with that. The intention is mainly to verify an opcode is usable in certain context.

> This also maybe mixes in with https://github.com/rust-bitcoin/rust-bitcoin/issues/1324

Yes, it does and my idea should work with it too. It'll be a bit more fiddly but not a huge issue, I think.

File: bitcoin/src/address.rs
Comment (apoelstra) [reply bitcoin/src/address.rs:330]: Ah, I see. agreed.

**Comment by apoelstra** [comment]
>Is your idea to just merge it nd iterate on that? That's not too bad for something you doubt can be done (I think it can).

Yes, that's my vote -- though I expect that whatever we release in 0.30 won't be the final form, and that we'll have more breakage in 0.31.

If you want to avoid repeated breakage like that, we should hold off on merging this until we've come up with a more comprehensive design. I'd rather just move forward and iterate, though I don't feel super strongly.

**Comment by Kixunil** [comment]
OK, no problem with merging and I will try to find some time to sketch up alternative before 0.30 is released. But even then I don't expect 0.30 to become stable. The whole stabilization project is an insane monstrosity and will take time.

**Comment by tcharding** [comment]
> The whole stabilization project is an insane monstrosity and will take time.

Lolz, the same thought had occurred to me of late.

**Comment by tcharding** [comment]
Mentioned in PR description but double posting, now includes importing with wildcard and removing the `opcodes::` path from any type starting with `OP_`.

**Comment by apoelstra** [comment]
Ok, weak concept ACK on this design. I agree it's annoying to write `op::OP_WHATEVER` but I also really dislike wildcard imports because it mean that when you see `OP_WHATEVER` you can't tell where it came from by searching the current file.

But for opcodes maybe it's okay to make an exception because there's so many of them and they're all easily identifiable.

**Review by apoelstra — DISMISSED** [review]
ACK d5d34a2a7611a6c3fe984df8397cde2b6567e386

**Comment by tcharding** [comment]
> But for opcodes maybe it's okay to make an exception because there's so many of them and they're all easily identifiable.

hmmm, if you have any apprehension over wildcard imports then perhaps this PR is not good because it allows use of naked `All`, I maintained usage of `opcodes::All` even in a file with wildcard import but this is not enforced by the compiler. I'm ambivalent, both `op::OP_FOO` and wildcard with `OP_FOO` seem ok to me. Perhaps we should add re-name of `opcodes` module to `op` back into this PR then users have the choice of doing either? (And for our code we can just use the module prefix by convention for any type that does not have `OP_` prefix.)

**Comment by Kixunil** [comment]
Oh, yes, accidental import of `All` is annoying. But maybe if it was named `AnyOpcode` instead it wouldn't be that bad.

One thing about wildcard imports: if you have only one it's easy to conclude "there's one wildcard import so the item must coms from it". It's still harder than not having it but the tradeoff seems in favor of using it here.

**Comment by tcharding** [comment]
Huh, hilariously the wildcard usage is currently supported in a super clean way _because we have the `all` submodule_, instead of removing it, just use it as is currently possible.

Excuse the noise, this was a very roundabout way to find a pretty basic improvement, amusingly no one else saw it either :)

**Comment by Kixunil** [comment]
Do we want to keep the name `all`?

**Comment by tcharding** [comment]
Yeah I rekon this usage by users of the lib reads nicely

```
use bitcoin::opcodes::all::*
use bitcoin::opcodes;
```

**Review by Kixunil — APPROVED** [review]
ACK c3e43995196757023b1dd35213d6b6c857713e29

I'd also ACK a future commit/PR explicitly promising in the doc that the `all` module will only ever contain opcodes prefixed with `OP_` so it's OK to glob-import as long as the user doesn't have any other `OP_` items in the module.

**Review by apoelstra — APPROVED** [review]
ACK c3e43995196757023b1dd35213d6b6c857713e29

**Pull request merged**
